File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/dominoes/dominoes.go
package dominoes

type Domino [2]int

func MakeChain(input []Domino) ([]Domino, bool) {
	if len(input) == 0 {
		return []Domino{}, true
	}

	if len(input) == 1 {
		d := input[0]
		if d[0] == d[1] {
			return []Domino{d}, true
		}
		return nil, false
	}

	for i, d := range input {
		for _, first := range []Domino{d, {d[1], d[0]}} {
			used := make([]bool, len(input))
			used[i] = true
			if chain, ok := backtrack(input, used, first[0], first[1], []Domino{first}, len(input)-1); ok {
				return chain, true
			}
		}
	}
	return nil, false
}

func backtrack(input []Domino, used []bool, start, end int, current []Domino, remaining int) ([]Domino, bool) {
	if remaining == 0 {
		if start == end {
			return current, true
		}
		return nil, false
	}

	for i, d := range input {
		if !used[i] {
			for _, rot := range []Domino{d, {d[1], d[0]}} {
				if rot[0] == end {
					newUsed := make([]bool, len(used))
					copy(newUsed, used)
					newUsed[i] = true
					newCurrent := append(current, rot)
					if chain, ok := backtrack(input, newUsed, start, rot[1], newCurrent, remaining-1); ok {
						return chain, true
					}
				}
			}
		}
	}
	return nil, false
}

File: src/2025-02-24-deepclaude-rust/java/exercises/practice/dominoes/src/main/java/Dominoes.java
import java.util.List;
import java.util.ArrayList;

class Dominoes {

    List<Domino> formChain(List<Domino> inputDominoes) throws ChainNotFoundException {
        if (inputDominoes.isEmpty()) {
            return new ArrayList<>();
        }

        List<Domino> result = new ArrayList<>();
        boolean[] used = new boolean[inputDominoes.size()];

        if (inputDominoes.size() == 1) {
            Domino domino = inputDominoes.get(0);
            if (domino.getLeft() != domino.getRight()) {
                throw new ChainNotFoundException("No chain found.");
            }
            return new ArrayList<>(inputDominoes);
        }

        for (int i = 0; i < inputDominoes.size(); i++) {
            Domino first = inputDominoes.get(i);
            // Try original orientation
            used[i] = true;
            result.add(first);
            if (backtrack(inputDominoes, used, result, first.getRight(), first.getLeft())) {
                return result;
            }
            result.remove(result.size() - 1);
            used[i] = false;

            // Try flipped orientation
            Domino flipped = new Domino(first.getRight(), first.getLeft());
            used[i] = true;
            result.add(flipped);
            if (backtrack(inputDominoes, used, result, flipped.getRight(), flipped.getLeft())) {
                return result;
            }
            result.remove(result.size() - 1);
            used[i] = false;
        }

        throw new ChainNotFoundException("No chain found.");
    }

    private boolean backtrack(List<Domino> dominoes, boolean[] used, List<Domino> current, int lastRight, int firstLeft) {
        if (current.size() == dominoes.size()) {
            return lastRight == firstLeft;
        }

        for (int i = 0; i < dominoes.size(); i++) {
            if (!used[i]) {
                Domino dom = dominoes.get(i);

                // Try original orientation
                if (dom.getLeft() == lastRight) {
                    used[i] = true;
                    current.add(dom);
                    if (backtrack(dominoes, used, current, dom.getRight(), firstLeft)) {
                        return true;
                    }
                    current.remove(current.size() - 1);
                    used[i] = false;
                }

                // Try flipped orientation
                if (dom.getRight() == lastRight) {
                    used[i] = true;
                    Domino flipped = new Domino(dom.getRight(), dom.getLeft());
                    current.add(flipped);
                    if (backtrack(dominoes, used, current, flipped.getRight(), firstLeft)) {
                        return true;
                    }
                    current.remove(current.size() - 1);
                    used[i] = false;
                }
            }
        }

        return false;
    }

}
